Yoast/wordpress-seo
# Qak 2350 implement wp remote handler

## Context


* Some servers don't have `curl_exec` available even though curl is installed. This causes Guzzle to fail while `wp_remote_request` would still work. This PR introduces a custom handler which uses `wp_remote_request` to do the final request.

## Summary


This PR can be summarized in the following changelog entry:

* Fixes a fatal error that would occur on sites that did have curl installed but did not have the PHP curl extension installed.

## Test instructions

This PR can be tested by following these steps:

* **The full SEMrush flow should be tested.**
* **When disabling the SEMrush feature the REST API should also be unavailable**.

(following test intr. added by @enricobattocchi)
### Reproduce the bug
This can be done with VVV:
* `vagrant up`
* `vagrant ssh`
* the following commands will probably need to be performed as `root` (use `sudo`)
* edit `/etc/php/7.2/fpm/conf.d/20-curl.ini`, comment out the line `extension=curl.so` (just put a `;` at the start)
* edit `/etc/php/7.2/fpm/php.ini`, add the line `allow_url_fopen=0` at the end
* exit the SSH shell
* reboot the VM: `vagrant halt`, then `vagrant up`
* check the PHP info (should be http://vvv.test/phpinfo/): there shouldn't be a `curl` section, and the value for `allow_url_fopen` should be set to "Off"
* Go to the dashboard of one of the WordPress sites and install the latest Yoast SEO plugin (15.1.2)
* Edit a new Post, see it throws an error:
```
 Fatal error: Uncaught Exception: GuzzleHttp requires cURL, the allow_url_fopen ini setting, or a custom HTTP handler.
in /srv/www/wordpress-one/public_html/wp-content/plugins/wordpress-seo/vendor_prefixed/guzzlehttp/guzzle/src/functions.php on line 105
```
* go to SEO→General, Integrations tab, disable SEMrush
* Edit a new post again and see the fatal error happens anyway.

## Test the fix
* **either** checkout the PR branch in Vagrant site plugin dir (should be something like `<vagran-dir>/www/wordpress-one/public_html/wp-content/plugins`) and build **or** create a zip from the PR branch (`grunt artifact`) and upload the ZIP rplacing the Yoast SEO plugin
* re-enable the SEMrush integration if you didn't do it yet
* Edit a post, see the error is not there anymore
* _Test the full SEMrush flow_, checking that in the console there are no errors related to the SEMrush endpoints
* disable the SEMrush integration
* visit a SEMrush endpoint, like http://one.wordpress.test/wp-json/yoast/v1/semrush/related_keyphrases and see the endpoint is now disabled: you get a response like this:
```
{"code":"rest_no_route","message":"No route was found matching the URL and request method","data":{"status":404}}
```
instead of
```
{"code":"rest_missing_callback_param","message":"Missing parameter(s): keyphrase, country_code","data":{"status":400,"params":["keyphrase","country_code"]}}
```


## UI changes

## Documentation

## Quality assurance


Fixes https://yoast.atlassian.net/browse/QAK-2350
